### PR TITLE
Converted ec2_metric_alarm to boto3. Added treat_missing_data option.

### PR DESCRIPTION
**Pull request by joshuah82**
##### SUMMARY

Converted the ec2_metric_alarm module to boto3. Added support for the Treat Missing Data option.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

ec2_metric_alarm module

##### ANSIBLE VERSION

```
ansible 2.2.0.0 (meredith cdec853e37) last updated 2016/11/15 15:12:03 (GMT -500)
  lib/ansible/modules/core: (detached HEAD fe9c56a003) last updated 2016/11/15 15:12:10 (GMT -500)
  lib/ansible/modules/extras: (detached HEAD f564e77a08) last updated 2016/11/15 15:12:10 (GMT -500)
  config file = 
  configured module search path = Default w/o overrides
```

##### ADDITIONAL INFORMATION

Configuration of the module is backwards compatible with the boto 2 version of the module. A warning is displayed when using symbols for the comparison option instead of the named ComparisonOperators supported in boto3.

```
 [WARNING]: Using the <=, <, > and >= operators has been deprecated. Please use LessThanOrEqualToThreshold, LessThanThreshold,
GreaterThanThreshold or GreaterThanOrEqualToThreshold instead.
```

### CODE REVIEW

**Comment by felipe1982** [comment]
I want to use this in my stable Ansible 2.3, but only need the **treat missing data as** part of it. Which parts do I need to just get only this functionality? I do not need the boto3 stuff, unless that is a requisite of the **treat missing data as** portions.

**Comment by felipe1982** [comment]
**update** fixed. PEBKAC

**Comment by sveneh** [comment]
I've added this module to my `library` folder using ansible 2.3.1.0. When I try to re-create my old alarms, I get the following error:

```
Traceback (most recent call last):
  File "/tmp/ansible_YJVC4C/ansible_module_ec2_metric_alarm.py", line 386, in <module>
    main()
  File "/tmp/ansible_YJVC4C/ansible_module_ec2_metric_alarm.py", line 377, in main
    create_metric_alarm(connection, module)
  File "/tmp/ansible_YJVC4C/ansible_module_ec2_metric_alarm.py", line 231, in create_metric_alarm
    InsufficientDataActions=insufficient_data_actions)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 310, in _api_call
    return self._make_api_call(operation_name, kwargs)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 573, in _make_api_call
    api_params, operation_model, context=request_context)
  File "/usr/local/lib/python2.7/dist-packages/botocore/client.py", line 628, in _convert_to_request_dict
    api_params, operation_model)
  File "/usr/local/lib/python2.7/dist-packages/botocore/validate.py", line 291, in serialize_to_request
    raise ParamValidationError(report=report.generate_report())
botocore.exceptions.ParamValidationError: Parameter validation failed:
Invalid type for parameter InsufficientDataActions, value: None, type: <type 'NoneType'>, valid types: <type 'list'>, <type 'tuple'>
```

as `insufficient_data_actions` is an optional parameter, I didn't set it in my playbook. I got it working by adding `insufficient_data_actions: []` into my playbook. But I don't think this is intended behaviour.

File: lib/ansible/modules/cloud/amazon/ec2_metric_alarm.py [hunk 1]
```diff
@@ -121,6 +132,17 @@
...
+        description:
+          - Sets how the alarm handles missing data points.
+        required: false
+        choices:
+          - 'breaching'
+          - 'notBreaching'
+          - 'ignore'
+          - 'missing'
+        default: 'missing'
+        version_added: "2.4"
```
Comment (gundalow) [inline lib/ansible/modules/cloud/amazon/ec2_metric_alarm.py:145]: 2.5

**Comment by torte** [comment]
Is there any chance this is getting rebased and merged on some point?

**Comment by willthames** [comment]
@torte my main objection is a lack of test suite (that's fair enough given how old this PR is, but makes it hard to merge a big change like this) - if there was a test suite I'd be happy to rebase and hope to get it into 2.8 - without a test suite I'd probably just aim for early in 2.9

If anyone reading this has a test suite for this module that they could contribute, that would definitely help get this over the line.

**Comment by mente** [comment]
> Added treat_missing_data option

isn't it the same provided currently by `insufficient_data_actions` option? If so, it should be replaced by `treat_missing_data` instead of adding on top.

Also meanwhile this is breaking change is it possible to rename the module? `ec2_metric_alarm` says nothing about cloudwatch :-/

**Comment by ibaldonl** [comment]
@mente treat_missing_data is not the same as insufficient_data_actions.
Also, this is not a breaking change, just an added option, no need to rename the module.

**Comment by tremble** [comment]
resolved_by_pr #62669

**Pull request closed without merging**